sphinx/domains/__init__.py

```diff
@@ -219,6 +226,17 @@ def __init__(self, env: "BuildEnvironment") -> None:
         self.objtypes_for_role = self._role2type.get    # type: Callable[[str], List[str]]
         self.role_for_objtype = self._type2role.get     # type: Callable[[str], str]
 
+    def setup(self) -> None:
+        """Set up domain object."""
+        from sphinx.domains.std import StandardDomain
+
+        # Add special hyperlink target for index pages (ex. modindex)
+        std = cast(StandardDomain, self.env.get_domain('std'))
+        for index in self.indices:
+            if index.name and index.localname:
+                docname = "%s-%s" % (self.name, index.name)
+                std.note_hyperlink_target(index.name, docname, '', index.localname)
```

Comment:
I wonder it would be better to prepend domain name to name of reference target (ex. ```:ref:`py-modindex` ``` or ```:ref:`py:modindex` ```).

Reply:
+1 for "prepend domain name" to avoid conflict.

Reply:
Thanks, updated.